app/AvailablePlugin/EmailAddressWidget/Controller/EmailAddressWidgetVerificationsController.php

```diff
@@ -0,0 +1,139 @@
+  // Class name, used by Cake
+  public $name = "EmailAddressWidgetVerifications";
+
+  public $uses = array(
+    'EmailAddressWidget.CoEmailAddressWidget',
+    'EmailAddressWidget.EmailAddressWidgetVerification',
+    'CoMessageTemplate'
+  );
+
+  /**
+   * Determine the CO ID based on some attribute of the request.
+   * This method is intended to be overridden by model-specific controllers.
+   *
+   * @since  COmanage Registry v4.1.0
+   * @return Integer CO ID, or null if not implemented or not applicable.
+   * @throws InvalidArgumentException
+   */
+
+  protected function calculateImpliedCoId($data = null) {
+    if(!empty($this->request->params['pass'][0])) {
+      $rec = $this->EmailAddressWidgetVerification->getRecordToVerify($this->request->params['pass'][0]);
+      if(isset($rec['CoEmailAddressWidget']["CoDashboardWidget"]["CoDashboard"]["co_id"])) {
+        return $rec['CoEmailAddressWidget']["CoDashboardWidget"]["CoDashboard"]["co_id"];
+      }
+    }
+
+    // Or try the default behavior
+    return parent::calculateImpliedCoId();
+  }
+
+  /**
+   * Step 2 in adding an email address:
+   * Verify the token associated with the new address.
+   *
+   * @since  COmanage Registry v4.1.0
+   */
+  public function verify($token) {
+    if(empty($token)) {
+      $this->Api->restResultHeader(HttpStatusCodesEnum::HTTP_BAD_REQUEST, _txt('er.token'));
+      return false;
+    }
+
+    // Retrieve the Verification record and the configuration
+    $rec = $this->EmailAddressWidgetVerification->getRecordToVerify($token);
+    if(empty($rec)) {
+      $this->Api->restResultHeader(HttpStatusCodesEnum::HTTP_NOT_FOUND, _txt('er.token'));
+      return false;
+    }
+    // Check if the person owns the records and can proceed with the verification
+    $CoPerson = ClassRegistry::init('CoPerson');
+    $actorCoPersonId = $CoPerson->idForIdentifier($this->cur_co["Co"]["id"], $this->Session->read('Auth.User.username'));
+
+    if($rec['EmailAddressWidgetVerification']['co_person_id'] != $actorCoPersonId) {
+      $this->Api->restResultHeader(HttpStatusCodesEnum::HTTP_NOT_FOUND,
+                                   _txt('er.cop.nf', array($this->Session->read('Auth.User.username'))));
+      return false;
+    }
+
+    try {
+      if(!$this->EmailAddressWidgetVerification->checkValidity($token)) {
+        $this->Api->restResultHeader(HttpStatusCodesEnum::HTTP_NOT_ACCEPTABLE, _txt('er.emailaddresswidget.timeout'));
+        return false;
+      }
+      $this->set('email_address_id', $this->EmailAddressWidgetVerification->addEmailToPerson($token, $actorCoPersonId));
+
+      $this->Api->restResultHeader(HttpStatusCodesEnum::HTTP_CREATED,
+                                   _txt('rs.added-a3', array($rec['EmailAddressWidgetVerification']['email'])));
+    } catch (Exception $e) {
+      $this->Api->restResultHeader(HttpStatusCodesEnum::HTTP_BAD_REQUEST, $e->getMessage());
+    }
```

Comment:
All of this logic should be moved into the model, which can throw an appropriate error (if needed) that the controller can catch and use to return back via ApiComponent.